# Exploring Hackers News Posts

In this project, we'll compare two different types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Introduction
First, we'll read in the data and remove the headers.

In [1]:
# Read in the data.
from csv import reader

open_file = open('hacker_news.csv')
read_file = reader(open_file)
hn = list(read_file)

hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing Headers from a List of Lists

In [2]:
# Isolate the headers.
headers = hn[0]
hn = hn[1:]
print(headers)
for row in hn[:5]:
    print(row)
    print('\n')


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']




We can see above that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. Let's start by exploring the number of comments for each type of post.

## Extracting Ask HN and Show HN Posts
First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = list()
show_posts = list()
other_posts = list()

for row in hn:
    title = row[1].lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print('Número de linhas:')
print(f'ask_posts = {len(ask_posts)}')
print(f'show_posts = {len(show_posts)}')
print(f'other_posts = {len(other_posts)}')

Número de linhas:
ask_posts = 1744
show_posts = 1162
other_posts = 17194


## Exploring the first five posts of Ask hn and Show hn
Below are the first five rows in the ask_posts and show_posts list of lists:

In [4]:
for row in ask_posts[:5]:
    print(row)
    print('\n')

['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55']


['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43']


['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']


['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20']


['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']




In [5]:
for row in show_posts[:5]:
    print(row)
    print('\n')

['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03']


['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46']


['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05']


['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11']


['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']




## Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [6]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for row in ask_posts:
    post_comments = int(row[4])
    total_ask_comments += post_comments

avg_ask_comments = total_ask_comments / len(ask_posts)
print(f'avg_ask_comments = {avg_ask_comments:.2f}')

avg_ask_comments = 14.04


In [7]:
# Calculate the average number of comments `Show HN` posts receive.
total_show_comments = 0

for row in show_posts:
    post_comments = int(row[4])
    total_show_comments += post_comments
    
avg_show_comments = total_show_comments / len(show_posts)
print(f'avg_show_comments = {avg_show_comments:.2f}')

avg_show_comments = 10.32


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [8]:
# Calculate the amount of ask posts created during each hour of the day and the number of comments received.
import datetime as dt

result_list = list()

for row in ask_posts:
    result_list.append([row[6], int(row[4])])


comments_by_hour = dict()
counts_by_hour = dict()
date_format = "%m/%d/%Y %H:%M" # According to our date column format

for row in result_list:
    date_str = row[0]
    comments = int(row[1])
    
    # Converting the date string to a datetime object and filter by hour
    datetime_object = dt.datetime.strptime(date_str, date_format)
    datetime_hour = datetime_object.hour
    
    if datetime_hour not in counts_by_hour:
        counts_by_hour[datetime_hour] = 1
        comments_by_hour[datetime_hour] = comments
    else:
        counts_by_hour[datetime_hour] += 1
        comments_by_hour[datetime_hour] += comments
    
for keys, values in sorted(comments_by_hour.items()):
    print(f'{keys} = {values}')

0 = 447
1 = 683
2 = 1381
3 = 421
4 = 337
5 = 464
6 = 397
7 = 267
8 = 492
9 = 251
10 = 793
11 = 641
12 = 687
13 = 1253
14 = 1416
15 = 4477
16 = 1814
17 = 1146
18 = 1439
19 = 1188
20 = 1722
21 = 1745
22 = 479
23 = 543


## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [9]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = list()

for keys, values in comments_by_hour.items():
    avg_by_hour.append([keys, values / counts_by_hour[keys]])

avg_by_hour

[[9, 5.5777777777777775],
 [13, 14.741176470588234],
 [10, 13.440677966101696],
 [14, 13.233644859813085],
 [16, 16.796296296296298],
 [23, 7.985294117647059],
 [12, 9.41095890410959],
 [17, 11.46],
 [15, 38.5948275862069],
 [21, 16.009174311926607],
 [20, 21.525],
 [2, 23.810344827586206],
 [18, 13.20183486238532],
 [3, 7.796296296296297],
 [5, 10.08695652173913],
 [19, 10.8],
 [1, 11.383333333333333],
 [22, 6.746478873239437],
 [8, 10.25],
 [4, 7.170212765957447],
 [0, 8.127272727272727],
 [6, 9.022727272727273],
 [7, 7.852941176470588],
 [11, 11.051724137931034]]

## Sorting and Printing Values from a List of Lists

In [10]:
from operator import itemgetter

avg_by_hour_sorted = sorted(avg_by_hour, key=itemgetter(1), reverse=True)

for hr, avg in avg_by_hour_sorted:
    print(f'{hr}H = {avg:.2f}')

15H = 38.59
2H = 23.81
20H = 21.52
16H = 16.80
21H = 16.01
13H = 14.74
10H = 13.44
14H = 13.23
18H = 13.20
17H = 11.46
1H = 11.38
11H = 11.05
19H = 10.80
8H = 10.25
5H = 10.09
12H = 9.41
6H = 9.02
0H = 8.13
23H = 7.99
7H = 7.85
3H = 7.80
4H = 7.17
22H = 6.75
9H = 5.58


In [11]:
# Sort the values and print the the 5 hours with the highest average comments.
print("Top 5 Hours for Ask Posts Comments")

for hr in avg_by_hour_sorted[:5]:
    datetime_object = dt.datetime.strptime(str(hr[0]), "%H")
    datetime_object_format = datetime_object.strftime("%H:%MH")
    
    print(f'{datetime_object_format} = {hr[1]:.2f} average comments per post')

Top 5 Hours for Ask Posts Comments
15:00H = 38.59 average comments per post
02:00H = 23.81 average comments per post
20:00H = 21.52 average comments per post
16:00H = 16.80 average comments per post
21:00H = 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est. That corresponds to 20:00 in Portugal timezone (UTC+00:00) Dublin, Edimburgo, Lisboa, Londres).

## Addition: Explore if show or ask posts receive more points on average

We can also explore if show or ask posts receive more points on average, by analyzing row 3.

In [12]:
ask_tot_num_points = 0
for row in ask_posts:
    num_points = int(row[3])
    ask_tot_num_points += num_points
    
ask_avg_num_points = ask_tot_num_points / len(ask_posts)
print(ask_avg_num_points)

15.061926605504587


In [13]:
show_tot_num_points = 0
for row in show_posts:
    num_points = int(row[3])
    show_tot_num_points += num_points
    
show_avg_num_points = show_tot_num_points / len(show_posts)
print(show_avg_num_points)

27.555077452667813


From above, we can see that show posts receive more points on average than ask posts (27.6 > 15.1). Therefore, let's focus our attention now on show posts.

## Finding the Amount of Show Posts and Points by Hour Created

In [14]:
show_counts_by_hour = dict()
show_points_by_hour = dict()

date_format = "%m/%d/%Y %H:%M"

for row in show_posts:
    time = row[-1]
    points = int(row[3])
    
    datetime_object = dt.datetime.strptime(time, date_format).hour
    
    if datetime_object not in show_counts_by_hour:
        show_counts_by_hour[datetime_object] = 1
        show_points_by_hour[datetime_object] = points
    else:
        show_counts_by_hour[datetime_object] += 1
        show_points_by_hour[datetime_object] += points

for keys, values in show_points_by_hour.items():
    print(f'{keys}H = {values}')

14H = 2187
22H = 1856
18H = 2215
7H = 494
20H = 1819
5H = 104
16H = 2634
19H = 1702
15H = 2228
3H = 679
17H = 2521
6H = 375
2H = 340
13H = 2438
8H = 519
21H = 866
4H = 386
11H = 1480
12H = 2543
23H = 1526
9H = 553
1H = 700
10H = 681
0H = 1173


## Calculating the Average Number of Points for Show HN Posts by Hour

In [15]:
# New empty dictionary to hold the new information
avg_show_points_hour = dict()

for hour, points in show_points_by_hour.items():
    avg_show_points_hour[hour] = points / show_counts_by_hour[hour]


# Sorting the new dictionary by descending order, according to the average points by hour
avg_show_points_hour_sorted = sorted(avg_show_points_hour.items(), key=itemgetter(1), reverse=True)

for hour, avg in avg_show_points_hour_sorted:
    print(f'{hour}:00H = {avg:.2f}')

23:00H = 42.39
12:00H = 41.69
22:00H = 40.35
0:00H = 37.84
18:00H = 36.31
11:00H = 33.64
19:00H = 30.95
20:00H = 30.32
15:00H = 28.56
16:00H = 28.32
17:00H = 27.11
14:00H = 25.43
3:00H = 25.15
1:00H = 25.00
13:00H = 24.63
6:00H = 23.44
7:00H = 19.00
10:00H = 18.92
9:00H = 18.43
21:00H = 18.43
8:00H = 15.26
4:00H = 14.85
2:00H = 11.33
5:00H = 5.47


Analyzing the results presented above, we can see that the hour that receives more points on average is 23:00H. Nevertheless, very similar results can be found at 12:00H and 22:00H. So, when considering the show points, the difference between the top 5 hours with higher average isn't as clear as the results we have seen for the ask posts average comments, where the second higher average hour had about 60% less comments than the first.

## Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.

In addition, we have also analyzed the average number points received by show posts by hour.